File: src/lib/vision/white-box-marking.ts
export interface WhiteBoxMark {
  number: number;
  x: number;
  y: number;
  width: number;
  height: number;
  area: number;
}

export interface SearchRegion {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface WhiteBoxMarkingInput {
  width: number;
  height: number;
  rgba: Uint8ClampedArray;
  whiteThreshold?: number | null;
  minAreaPx?: number | null;
  searchRegion?: SearchRegion;
}

export interface WhiteBoxMarkingResult {
  width: number;
  height: number;
  rgba: Uint8ClampedArray;
  boxes: WhiteBoxMark[];
}

const RGBA_STRIDE = 4;
const RED_OFFSET = 0;
const GREEN_OFFSET = 1;
const BLUE_OFFSET = 2;
const ALPHA_OFFSET = 3;
const OPAQUE_ALPHA = 255;
const LUMA_RED_WEIGHT = 0.299;
const LUMA_GREEN_WEIGHT = 0.587;
const LUMA_BLUE_WEIGHT = 0.114;
const TWO_BIT_GRAY_LEVELS = [0, 85, 170, 255] as const;
const TWO_BIT_GRAY_BREAKS = [64, 128, 192] as const;
const AUTO_THRESHOLD_NO_MATCH = 256;
const MIN_COMPONENT_AREA_RATIO = 0.00008;
const MAX_MARKING_SIZE_RATIO = 0.85;
const BOX_BORDER_RGBA = [255, 0, 0, 255] as const;
const LABEL_BACKGROUND_RGBA = [255, 255, 255, 255] as const;
const LABEL_TEXT_RGBA = [0, 0, 0, 255] as const;
const DIGIT_WIDTH_PX = 3;
const DIGIT_HEIGHT_PX = 5;
const DIGIT_GAP_PX = 1;
const LABEL_PADDING_PX = 1;

// Tiny bitmap font used to draw stable box numbers directly onto the canvas pixels.
const DIGITS: Record<string, readonly string[]> = {
  "0": ["111", "101", "101", "101", "111"],
  "1": ["010", "110", "010", "010", "111"],
  "2": ["111", "001", "111", "100", "111"],
  "3": ["111", "001", "111", "001", "111"],
  "4": ["101", "101", "111", "001", "001"],
  "5": ["111", "100", "111", "001", "111"],
  "6": ["111", "100", "111", "101", "111"],
  "7": ["111", "001", "010", "010", "010"],
  "8": ["111", "101", "111", "101", "111"],
  "9": ["111", "101", "111", "001", "111"],
};

type Component = { x0: number; y0: number; x1: number; y1: number; area: number };

export function markWhiteBoxes(input: WhiteBoxMarkingInput): WhiteBoxMarkingResult {
  assertRgba(input);
  const hasThreshold = input.whiteThreshold !== undefined && input.whiteThreshold !== null;
  const gray = hasThreshold ? toContinuousGray(input.rgba) : toTwoBitGray(input.rgba);
  const boxes = findWhiteBoxes(gray, input.width, input.height, input.whiteThreshold, input.minAreaPx, input.searchRegion);

  return renderWhiteBoxMarks(input.width, input.height, gray, boxes);
}

export function renderWhiteBoxMarks(
  width: number,
  height: number,
  gray: Uint8ClampedArray,
  boxes: WhiteBoxMark[],
): WhiteBoxMarkingResult {
  const marked = grayToRgba(gray);
  drawBoxes(marked, width, height, boxes);

  return { width, height, rgba: marked, boxes };
}

export function toTwoBitGrayRgba(input: WhiteBoxMarkingInput): Uint8ClampedArray {
  assertRgba(input);

  return grayToRgba(toTwoBitGray(input.rgba));
}

export function toThresholdPreviewRgba(input: WhiteBoxMarkingInput, whiteThreshold: number): Uint8ClampedArray {
  assertRgba(input);
  const gray = toContinuousGray(input.rgba);
  const threshold = clampByte(whiteThreshold);

  for (let index = 0; index < gray.length; index += 1) {
    if (gray[index] >= threshold) {
      gray[index] = 255;
    }
  }

  return grayToRgba(gray);
}

export function renderSelectedWhiteBoxMarks(
  input: WhiteBoxMarkingInput,
  boxes: WhiteBoxMark[],
  whiteThreshold?: number,
): WhiteBoxMarkingResult {
  assertRgba(input);
  const rgba =
    whiteThreshold === undefined
      ? grayToRgba(toTwoBitGray(input.rgba))
      : toThresholdPreviewRgba(input, whiteThreshold);
  drawBoxes(rgba, input.width, input.height, boxes);

  return { width: input.width, height: input.height, rgba, boxes };
}

export function rgbaToBase64(rgba: Uint8ClampedArray): string {
  let binary = "";
  const chunk = 8192;
  for (let index = 0; index < rgba.length; index += chunk) {
    binary += String.fromCharCode(...rgba.slice(index, index + chunk));
  }
  return btoa(binary);
}

export function base64ToRgba(value: string): Uint8ClampedArray {
  const binary = atob(value);
  const out = new Uint8ClampedArray(binary.length);
  for (let index = 0; index < binary.length; index += 1) out[index] = binary.charCodeAt(index);
  return out;
}

export function drawRgbaToCanvas(
  canvas: HTMLCanvasElement,
  width: number,
  height: number,
  rgba: Uint8ClampedArray,
): void {
  canvas.width = width;
  canvas.height = height;
  const context = canvas.getContext("2d");
  if (context === null) throw new Error("2d canvas context unavailable");
  context.putImageData(new ImageData(new Uint8ClampedArray(rgba), width, height), 0, 0);
}

export async function readImageFile(file: File): Promise<WhiteBoxMarkingInput> {
  const bitmap = await createImageBitmap(file);
  const canvas = new OffscreenCanvas(bitmap.width, bitmap.height);
  const context = canvas.getContext("2d");
  if (context === null) throw new Error("2d canvas context unavailable");
  context.drawImage(bitmap, 0, 0);
  const image = context.getImageData(0, 0, bitmap.width, bitmap.height);
  return { width: bitmap.width, height: bitmap.height, rgba: image.data };
}

function assertRgba(input: WhiteBoxMarkingInput): void {
  const expected = input.width * input.height * RGBA_STRIDE;
  if (input.width <= 0 || input.height <= 0) throw new Error("image dimensions must be positive");
  if (input.rgba.length !== expected) throw new Error(`rgba length mismatch: ${input.rgba.length}`);
}

export function toContinuousGray(rgba: Uint8ClampedArray): Uint8ClampedArray {
  const gray = new Uint8ClampedArray(rgba.length / RGBA_STRIDE);

  for (let index = 0; index < rgba.length; index += RGBA_STRIDE) {
    const lum =
      LUMA_RED_WEIGHT * rgba[index + RED_OFFSET] +
      LUMA_GREEN_WEIGHT * rgba[index + GREEN_OFFSET] +
      LUMA_BLUE_WEIGHT * rgba[index + BLUE_OFFSET];

    gray[index / RGBA_STRIDE] = clampByte(lum);
  }

  return gray;
}

function toTwoBitGray(rgba: Uint8ClampedArray): Uint8ClampedArray {
  const gray = new Uint8ClampedArray(rgba.length / RGBA_STRIDE);
  for (let index = 0; index < rgba.length; index += RGBA_STRIDE) {
    const lum =
      LUMA_RED_WEIGHT * rgba[index + RED_OFFSET] +
      LUMA_GREEN_WEIGHT * rgba[index + GREEN_OFFSET] +
      LUMA_BLUE_WEIGHT * rgba[index + BLUE_OFFSET];
    gray[index / RGBA_STRIDE] = quantize(lum);
  }
  return gray;
}

function quantize(value: number): number {
  if (value < TWO_BIT_GRAY_BREAKS[0]) return TWO_BIT_GRAY_LEVELS[0];
  if (value < TWO_BIT_GRAY_BREAKS[1]) return TWO_BIT_GRAY_LEVELS[1];
  if (value < TWO_BIT_GRAY_BREAKS[2]) return TWO_BIT_GRAY_LEVELS[2];
  return TWO_BIT_GRAY_LEVELS[3];
}

function clampByte(value: number): number {
  return Math.max(0, Math.min(255, Math.round(value)));
}

function findWhiteBoxes(
  gray: Uint8ClampedArray,
  width: number,
  height: number,
  whiteThreshold: number | null | undefined,
  minAreaPx: number | null | undefined,
  searchRegion: SearchRegion | undefined,
): WhiteBoxMark[] {
  const seen = new Uint8Array(width * height);
  const boxes: WhiteBoxMark[] = [];
  const region = clampRegion(width, height, searchRegion);
  if (region.width <= 0 || region.height <= 0) return [];
  const threshold = whiteThreshold ?? autoThreshold(gray, width, region);
  const minArea = minAreaPx ?? autoMinArea(region);
  for (let y = region.y; y < region.y + region.height; y += 1) {
    for (let x = region.x; x < region.x + region.width; x += 1) {
      const index = y * width + x;
      if (gray[index] < threshold || seen[index] === 1) continue;
      const component = walkComponent(gray, seen, width, height, index, threshold, region);
      if (component.area >= minArea && looksLikeMarking(component, region)) boxes.push(toBox(boxes.length + 1, component));
    }
  }
  return boxes.sort(readingOrder).map((box, index) => ({ ...box, number: index + 1 }));
}

function autoThreshold(gray: Uint8ClampedArray, width: number, region: SearchRegion): number {
  const counts: Record<number, number> = Object.fromEntries(TWO_BIT_GRAY_LEVELS.map((level) => [level, 0]));
  for (let y = region.y; y < region.y + region.height; y += 1) {
    for (let x = region.x; x < region.x + region.width; x += 1) counts[gray[y * width + x]] += 1;
  }
  const background = TWO_BIT_GRAY_LEVELS.reduce((best, level) => (counts[level] > counts[best] ? level : best), 0);
  return TWO_BIT_GRAY_LEVELS.find((level) => level > background && counts[level] > 0) ?? AUTO_THRESHOLD_NO_MATCH;
}

function autoMinArea(region: SearchRegion): number {
  return Math.max(2, Math.round(region.width * region.height * MIN_COMPONENT_AREA_RATIO));
}

function looksLikeMarking(component: Component, region: SearchRegion): boolean {
  if (component.x1 <= component.x0 || component.y1 <= component.y0) return false;
  const componentWidth = component.x1 - component.x0 + 1;
  const componentHeight = component.y1 - component.y0 + 1;
  return componentWidth < region.width * MAX_MARKING_SIZE_RATIO && componentHeight < region.height * MAX_MARKING_SIZE_RATIO;
}

function clampRegion(width: number, height: number, region: SearchRegion | undefined): SearchRegion {
  if (region === undefined) return { x: 0, y: 0, width, height };
  const x = Math.max(0, Math.min(width, Math.floor(region.x)));
  const y = Math.max(0, Math.min(height, Math.floor(region.y)));
  const x2 = Math.max(0, Math.min(width, Math.ceil(region.x + region.width)));
  const y2 = Math.max(0, Math.min(height, Math.ceil(region.y + region.height)));
  return { x, y, width: Math.max(0, x2 - x), height: Math.max(0, y2 - y) };
}

function walkComponent(
  gray: Uint8ClampedArray,
  seen: Uint8Array,
  width: number,
  height: number,
  start: number,
  whiteThreshold: number,
  region: SearchRegion,
): Component {
  const stack = [start];
  let x0 = start % width;
  let x1 = x0;
  let y0 = Math.floor(start / width);
  let y1 = y0;
  let area = 0;
  seen[start] = 1;
  while (stack.length > 0) {
    const current = stack.pop() as number;
    const x = current % width;
    const y = Math.floor(current / width);
    area += 1;
    x0 = Math.min(x0, x);
    y0 = Math.min(y0, y);
    x1 = Math.max(x1, x);
    y1 = Math.max(y1, y);
    pushNeighbors(gray, seen, stack, width, height, current, whiteThreshold, region);
  }
  return { x0, y0, x1, y1, area };
}

function pushNeighbors(
  gray: Uint8ClampedArray,
  seen: Uint8Array,
  stack: number[],
  width: number,
  height: number,
  current: number,
  threshold: number,
  region: SearchRegion,
): void {
  const x = current % width;
  const y = Math.floor(current / width);
  for (let nx = x - 1; nx <= x + 1; nx += 1) {
    for (let ny = y - 1; ny <= y + 1; ny += 1) {
      if (nx === x && ny === y) continue;
      if (nx < region.x || ny < region.y || nx >= region.x + region.width || ny >= region.y + region.height) continue;
      const ni = ny * width + nx;
      if (seen[ni] === 0 && gray[ni] >= threshold) {
        seen[ni] = 1;
        stack.push(ni);
      }
    }
  }
}

function toBox(number: number, component: Component): WhiteBoxMark {
  return {
    number,
    x: component.x0,
    y: component.y0,
    width: component.x1 - component.x0 + 1,
    height: component.y1 - component.y0 + 1,
    area: component.area,
  };
}

function readingOrder(a: WhiteBoxMark, b: WhiteBoxMark): number {
  return a.y === b.y ? a.x - b.x : a.y - b.y;
}

function grayToRgba(gray: Uint8ClampedArray): Uint8ClampedArray {
  const out = new Uint8ClampedArray(gray.length * RGBA_STRIDE);
  for (let index = 0; index < gray.length; index += 1) {
    const pos = index * RGBA_STRIDE;
    out[pos + RED_OFFSET] = gray[index];
    out[pos + GREEN_OFFSET] = gray[index];
    out[pos + BLUE_OFFSET] = gray[index];
    out[pos + ALPHA_OFFSET] = OPAQUE_ALPHA;
  }
  return out;
}

function drawBoxes(rgba: Uint8ClampedArray, width: number, height: number, boxes: WhiteBoxMark[]): void {
  for (const box of boxes) {
    drawRect(rgba, width, height, box);
    drawLabel(rgba, width, height, box);
  }
}

function drawRect(rgba: Uint8ClampedArray, width: number, height: number, box: WhiteBoxMark): void {
  const x2 = box.x + box.width - 1;
  const y2 = box.y + box.height - 1;
  for (let x = box.x; x <= x2; x += 1) {
    setPixel(rgba, width, height, x, box.y, BOX_BORDER_RGBA);
    setPixel(rgba, width, height, x, y2, BOX_BORDER_RGBA);
  }
  for (let y = box.y; y <= y2; y += 1) {
    setPixel(rgba, width, height, box.x, y, BOX_BORDER_RGBA);
    setPixel(rgba, width, height, x2, y, BOX_BORDER_RGBA);
  }
}

function drawLabel(rgba: Uint8ClampedArray, width: number, height: number, box: WhiteBoxMark): void {
  const text = String(box.number);
  const labelWidth = text.length * (DIGIT_WIDTH_PX + DIGIT_GAP_PX) + LABEL_PADDING_PX * 2;
  const labelHeight = DIGIT_HEIGHT_PX + LABEL_PADDING_PX * 2;
  fillRect(rgba, width, height, box.x, box.y, labelWidth, labelHeight, LABEL_BACKGROUND_RGBA);
  [...text].forEach((digit, offset) =>
    drawDigit(
      rgba,
      width,
      height,
      box.x + LABEL_PADDING_PX + offset * (DIGIT_WIDTH_PX + DIGIT_GAP_PX),
      box.y + LABEL_PADDING_PX,
      digit,
    ),
  );
}

function drawDigit(
  rgba: Uint8ClampedArray,
  width: number,
  height: number,
  x: number,
  y: number,
  digit: string,
): void {
  DIGITS[digit]?.forEach((bits, row) => {
    [...bits].forEach((bit, col) => {
      if (bit === "1") setPixel(rgba, width, height, x + col, y + row, LABEL_TEXT_RGBA);
    });
  });
}

function fillRect(
  rgba: Uint8ClampedArray,
  width: number,
  height: number,
  x: number,
  y: number,
  w: number,
  h: number,
  color: readonly number[],
): void {
  for (let yy = y; yy < y + h; yy += 1) {
    for (let xx = x; xx < x + w; xx += 1) setPixel(rgba, width, height, xx, yy, color);
  }
}

function setPixel(
  rgba: Uint8ClampedArray,
  width: number,
  height: number,
  x: number,
  y: number,
  color: readonly number[],
): void {
  if (x < 0 || y < 0 || x >= width || y >= height) return;
  const pos = (y * width + x) * RGBA_STRIDE;
  rgba[pos + RED_OFFSET] = color[RED_OFFSET];
  rgba[pos + GREEN_OFFSET] = color[GREEN_OFFSET];
  rgba[pos + BLUE_OFFSET] = color[BLUE_OFFSET];
  rgba[pos + ALPHA_OFFSET] = color[ALPHA_OFFSET];
}

File: BE/app/domain/white_box_marking.py
"""2-bit grayscale conversion and white-region box marking."""

from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Final

from BE.errors.apperror import AppError
from BE.errors.codes import ErrorCode

MIN_PIXEL_INTENSITY: Final[int] = 0
MAX_PIXEL_INTENSITY: Final[int] = 255
WHITE_PIXEL_VALUE: Final[int] = 255
BYTES_PER_RGBA_PIXEL: Final[int] = 4

RED_COLOR_RGBA: Final[tuple[int, int, int, int]] = (255, 0, 0, 255)
WHITE_COLOR_RGBA: Final[tuple[int, int, int, int]] = (255, 255, 255, 255)
BLACK_COLOR_RGBA: Final[tuple[int, int, int, int]] = (0, 0, 0, 255)



@dataclass(frozen=True)
class SearchRegion:
    x: int
    y: int
    width: int
    height: int


@dataclass(frozen=True)
class MarkingOptions:
    white_threshold: int | None = None
    min_area_px: int | None = None
    search_region: SearchRegion | None = None


@dataclass(frozen=True)
class MarkedBox:
    number: int
    x: int
    y: int
    width: int
    height: int
    area: int


@dataclass(frozen=True)
class MarkingResult:
    width: int
    height: int
    rgba_base64: str
    boxes: tuple[MarkedBox, ...]


_DIGITS: dict[str, tuple[str, ...]] = {
    "0": ("111", "101", "101", "101", "111"),
    "1": ("010", "110", "010", "010", "111"),
    "2": ("111", "001", "111", "100", "111"),
    "3": ("111", "001", "111", "001", "111"),
    "4": ("101", "101", "111", "001", "001"),
    "5": ("111", "100", "111", "001", "111"),
    "6": ("111", "100", "111", "101", "111"),
    "7": ("111", "001", "010", "010", "010"),
    "8": ("111", "101", "111", "101", "111"),
    "9": ("111", "101", "111", "001", "111"),
}


def mark_white_boxes(
    *,
    width: int,
    height: int,
    rgba_base64: str,
    options: MarkingOptions | None = None,
) -> MarkingResult:
    opts = options or MarkingOptions()
    rgba = _decode_rgba(width, height, rgba_base64)
    gray = _to_two_bit_gray(rgba)
    boxes = _find_white_boxes(gray, width, height, opts)
    preview = _threshold_preview(gray, opts.white_threshold) if opts.white_threshold is not None else gray
    marked = bytearray(_gray_to_rgba(preview))
    _draw_boxes(marked, width, height, boxes)
    encoded = base64.b64encode(bytes(marked)).decode("ascii")
    return MarkingResult(width=width, height=height, rgba_base64=encoded, boxes=tuple(boxes))


def _decode_rgba(width: int, height: int, rgba_base64: str) -> bytes:
    if width <= 0 or height <= 0:
        _bad("width and height must be positive", {"Width": width, "Height": height})
    try:
        raw = base64.b64decode(rgba_base64, validate=True)
    except ValueError as exc:
        _bad("RgbaBase64 must be valid base64", {"Cause": str(exc)})
    expected = width * height * 4
    if len(raw) != expected:
        _bad("RgbaBase64 length does not match dimensions", {"Expected": expected, "Actual": len(raw)})
    return raw


def _bad(message: str, details: dict[str, object]) -> None:
    raise AppError(ErrorCode.E_BE_BAD_REQUEST, message, details)


def _to_two_bit_gray(rgba: bytes) -> bytearray:
    out = bytearray(len(rgba) // 4)
    for idx in range(0, len(rgba), 4):
        lum = int(0.299 * rgba[idx] + 0.587 * rgba[idx + 1] + 0.114 * rgba[idx + 2])
        out[idx // 4] = _quantize(lum)
    return out


def _quantize(value: int) -> int:
    if value < 64:
        return 0
    if value < 128:
        return 85
    if value < 192:
        return 170
    return 255


def clamp_pixel_intensity(raw_threshold: float | int) -> int:
    """Safely rounds and clamps an arbitrary threshold into valid 8-bit range [0, 255]."""
    rounded_threshold = round(raw_threshold)
    return max(
        MIN_PIXEL_INTENSITY,
        min(MAX_PIXEL_INTENSITY, rounded_threshold),
    )


def build_threshold_lookup_table(clamped_threshold: int) -> bytes:
    """Precompute 256-byte lookup table (LUT) eliminating per-pixel branching."""
    return bytes(
        WHITE_PIXEL_VALUE if value >= clamped_threshold else value
        for value in range(MAX_PIXEL_INTENSITY + 1)
    )


def _threshold_preview(
    gray: bytearray | bytes,
    white_threshold: float | int,
) -> bytearray:
    """Highlight pixels at or above `white_threshold` by setting them to pure white."""
    clamped_threshold = clamp_pixel_intensity(white_threshold)
    lookup_table = build_threshold_lookup_table(clamped_threshold)
    return bytearray(gray.translate(lookup_table))


# Public alias adhering to Prompt 26
threshold_preview = _threshold_preview



def _find_white_boxes(
    gray: bytearray,
    width: int,
    height: int,
    options: MarkingOptions,
) -> list[MarkedBox]:
    seen = bytearray(width * height)
    boxes: list[MarkedBox] = []
    region = _clamp_region(width, height, options.search_region)
    if region.width <= 0 or region.height <= 0:
        return []
    threshold = options.white_threshold if options.white_threshold is not None else _auto_threshold(gray, width, region)
    min_area = options.min_area_px if options.min_area_px is not None else _auto_min_area(region)
    for y in range(region.y, region.y + region.height):
        for x in range(region.x, region.x + region.width):
            index = y * width + x
            value = gray[index]
            if value < threshold or seen[index] == 1:
                continue
            component = _walk_component(gray, seen, width, height, index, threshold, region)
            if component.area >= min_area and _looks_like_marking(component, region):
                boxes.append(_numbered_box(len(boxes) + 1, component))
    boxes.sort(key=lambda box: (box.y, box.x))
    return [_renumber(i + 1, box) for i, box in enumerate(boxes)]


def _auto_threshold(gray: bytearray, width: int, region: SearchRegion) -> int:
    counts = {0: 0, 85: 0, 170: 0, 255: 0}
    for y in range(region.y, region.y + region.height):
        start = y * width + region.x
        end = start + region.width
        for value in gray[start:end]:
            counts[value] += 1
    background = max(counts, key=lambda level: counts[level])
    for level in (85, 170, 255):
        if level > background and counts[level] > 0:
            return level
    return 256


def _auto_min_area(region: SearchRegion) -> int:
    return max(2, round(region.width * region.height * 0.00008))


@dataclass(frozen=True)
class _Component:
    x0: int
    y0: int
    x1: int
    y1: int
    area: int


def _looks_like_marking(component: _Component, region: SearchRegion) -> bool:
    if component.x1 <= component.x0 or component.y1 <= component.y0:
        return False
    component_width = component.x1 - component.x0 + 1
    component_height = component.y1 - component.y0 + 1
    return component_width < region.width * 0.85 and component_height < region.height * 0.85


def _clamp_region(width: int, height: int, region: SearchRegion | None) -> SearchRegion:
    if region is None:
        return SearchRegion(x=0, y=0, width=width, height=height)
    x = max(0, min(width, region.x))
    y = max(0, min(height, region.y))
    x2 = max(0, min(width, region.x + region.width))
    y2 = max(0, min(height, region.y + region.height))
    return SearchRegion(x=x, y=y, width=max(0, x2 - x), height=max(0, y2 - y))


def _walk_component(
    gray: bytearray,
    seen: bytearray,
    width: int,
    height: int,
    start: int,
    threshold: int,
    region: SearchRegion,
) -> _Component:
    stack = [start]
    seen[start] = 1
    x0 = x1 = start % width
    y0 = y1 = start // width
    area = 0
    while stack:
        current = stack.pop()
        area += 1
        x = current % width
        y = current // width
        x0, y0, x1, y1 = min(x0, x), min(y0, y), max(x1, x), max(y1, y)
        _push_neighbors(gray, seen, stack, width, height, current, threshold, region)
    return _Component(x0=x0, y0=y0, x1=x1, y1=y1, area=area)


def _push_neighbors(
    gray: bytearray,
    seen: bytearray,
    stack: list[int],
    width: int,
    height: int,
    current: int,
    threshold: int,
    region: SearchRegion,
) -> None:
    x = current % width
    y = current // width
    for nx in range(x - 1, x + 2):
        for ny in range(y - 1, y + 2):
            if nx == x and ny == y:
                continue
            if (
                nx < region.x
                or ny < region.y
                or nx >= region.x + region.width
                or ny >= region.y + region.height
                or nx >= width
                or ny >= height
            ):
                continue
            ni = ny * width + nx
            if seen[ni] == 0 and gray[ni] >= threshold:
                seen[ni] = 1
                stack.append(ni)


def _numbered_box(number: int, component: _Component) -> MarkedBox:
    return MarkedBox(
        number=number,
        x=component.x0,
        y=component.y0,
        width=component.x1 - component.x0 + 1,
        height=component.y1 - component.y0 + 1,
        area=component.area,
    )


def _renumber(number: int, box: MarkedBox) -> MarkedBox:
    return MarkedBox(number=number, x=box.x, y=box.y, width=box.width, height=box.height, area=box.area)


def _gray_to_rgba(gray: bytearray) -> bytes:
    out = bytearray(len(gray) * BYTES_PER_RGBA_PIXEL)
    for idx, value in enumerate(gray):
        pos = idx * BYTES_PER_RGBA_PIXEL
        out[pos : pos + BYTES_PER_RGBA_PIXEL] = bytes((value, value, value, WHITE_PIXEL_VALUE))
    return bytes(out)


def _draw_boxes(rgba: bytearray, width: int, height: int, boxes: list[MarkedBox]) -> None:
    for box in boxes:
        _draw_rect(rgba, width, height, box)
        _draw_label(rgba, width, height, box)


def _draw_rect(rgba: bytearray, width: int, height: int, box: MarkedBox) -> None:
    x2 = box.x + box.width - 1
    y2 = box.y + box.height - 1
    for x in range(box.x, x2 + 1):
        _set_pixel(rgba, width, height, x, box.y, RED_COLOR_RGBA)
        _set_pixel(rgba, width, height, x, y2, RED_COLOR_RGBA)
    for y in range(box.y, y2 + 1):
        _set_pixel(rgba, width, height, box.x, y, RED_COLOR_RGBA)
        _set_pixel(rgba, width, height, x2, y, RED_COLOR_RGBA)


def _draw_label(rgba: bytearray, width: int, height: int, box: MarkedBox) -> None:
    text = str(box.number)
    _fill_rect(rgba, width, height, box.x, box.y, len(text) * 4 + 2, 7, WHITE_COLOR_RGBA)
    for offset, digit in enumerate(text):
        _draw_digit(rgba, width, height, box.x + 1 + offset * 4, box.y + 1, digit)


def _draw_digit(rgba: bytearray, width: int, height: int, x: int, y: int, digit: str) -> None:
    for row, bits in enumerate(_DIGITS.get(digit, ())):
        for col, bit in enumerate(bits):
            if bit == "1":
                _set_pixel(rgba, width, height, x + col, y + row, BLACK_COLOR_RGBA)


def _fill_rect(
    rgba: bytearray,
    width: int,
    height: int,
    x: int,
    y: int,
    w: int,
    h: int,
    color: tuple[int, int, int, int],
) -> None:
    for yy in range(y, y + h):
        for xx in range(x, x + w):
            _set_pixel(rgba, width, height, xx, yy, color)


def _set_pixel(
    rgba: bytearray,
    width: int,
    height: int,
    x: int,
    y: int,
    color: tuple[int, int, int, int],
) -> None:
    if x < 0 or y < 0 or x >= width or y >= height:
        return
    pos = (y * width + x) * BYTES_PER_RGBA_PIXEL
    rgba[pos : pos + BYTES_PER_RGBA_PIXEL] = bytes(color)

